### PR TITLE
etcdserver: ignore old leader's request to revoke lease

### PR DESCRIPTION
**Pull request by tangcong**

### CODE REVIEW

**Comment by tangcong** [comment]
Fix #12528 (old leader still revokes lease after it steped to follower if cpu or disk io latency is high) @jpbetz @gyuho

File: server/etcdserver/server.go [hunk 1]
```diff
@@ -1051,6 +1051,19 @@ func (s *EtcdServer) run() {
...
 					s.GoAttach(func() {
+						// If the cpu and disk io latency is high, the old leader may still revoke the lease after becoming the follower.
+						// the root cause is that etcdserver processing log entries will be blocked due to high cpu and disk io latency, which makes it impossible to clear the lease and update the membership in time.
+						// this PR solves this issue by get the leader identity directly from the raft state machine.
+						// more detail info please see issue https://github.com/etcd-io/etcd/issues/12528
+						if !s.isLeader() || !s.isLeaderInRaftStateMachine() {
+							lg.Warn(
+								"ignore older leader revoke lease request",
+								zap.String("lease-id", fmt.Sprintf("%016x", lease.ID)),
+								zap.Bool("isLeader", s.isLeader()),
```
Comment (gyuho) [inline server/etcdserver/server.go:1062]: Can we be more consistent with casing? `is-leader`?

File: server/etcdserver/server.go [hunk 2]
```diff
@@ -1288,6 +1301,15 @@ func (s *EtcdServer) isLeader() bool {
 	return uint64(s.ID()) == s.Lead()
 }
 
+func (s *EtcdServer) isLeaderInRaftStateMachine() bool {
+	rs := s.raftStatus()
+	// leader's raftStatus.Progress is not nil
```
Comment (gyuho) [inline server/etcdserver/server.go:1306]: We can just `return rs.Progress != nil`?

File: server/etcdserver/server.go [hunk 3]
```diff
@@ -1051,6 +1051,19 @@ func (s *EtcdServer) run() {
 					}
 					lid := lease.ID
 					s.GoAttach(func() {
+						// If the cpu and disk io latency is high, the old leader may still revoke the lease after becoming the follower.
+						// the root cause is that etcdserver processing log entries will be blocked due to high cpu and disk io latency, which makes it impossible to clear the lease and update the membership in time.
+						// this PR solves this issue by get the leader identity directly from the raft state machine.
+						// more detail info please see issue https://github.com/etcd-io/etcd/issues/12528
+						if !s.isLeader() || !s.isLeaderInRaftStateMachine() {
```
Comment (gyuho) [inline server/etcdserver/server.go:1058]: Can we explain why `s.isLeader()` not enough?

File: server/etcdserver/server.go [hunk 4]
```diff
@@ -1288,6 +1301,15 @@ func (s *EtcdServer) isLeader() bool {
 	return uint64(s.ID()) == s.Lead()
 }
 
+func (s *EtcdServer) isLeaderInRaftStateMachine() bool {
+	rs := s.raftStatus()
```
Comment (gyuho) [inline server/etcdserver/server.go:1305]: Also, can we explain why querying `raftStatus` is thread safe?

File: server/etcdserver/server.go (hunk 1 above)
Comment (tangcong) [reply server/etcdserver/server.go]: updated. thanks.

File: server/etcdserver/server.go (hunk 2 above)
Comment (tangcong) [reply server/etcdserver/server.go:1309]: updated.

File: server/etcdserver/server.go (hunk 3 above)
Comment (tangcong) [reply server/etcdserver/server.go:1059]: updated.

File: server/etcdserver/server.go (hunk 4 above)
Comment (tangcong) [reply server/etcdserver/server.go:1308]: updated.

**Comment by tangcong** [comment]
updated. thanks. @gyuho  semaphores ci is failed, It seems that the failed test case is not related to this PR. my local e2e test is ok.

**Review by jiapeish — APPROVED** [review]
+1. We've verified this PR and it takes effect.

**Comment by jiapeish** [comment]
I noticed this PR is here for almost 1 month... Are there any more reviewers?

**Review by ptabor — COMMENTED** [review]
1. PRs that are not green on the list of PRs have lower chances to get attention. 
I know its due to test flakiness, but still recommend running 'git commit --amend; git push' to trigger the re-test. 

2. It's hard here, but it would be good to have a test-case (at least a unit one) that shows this scenario. 

3. I'm new to this code, but it seems unnatural that the decision to expire lease is not just propagated by RAFT... Alternatively the decision to revoke lease might contains term and its only executed (accepted by current leader in ) if the senders term matches the current term. For my education: were this option considered and refused for some reasons ? 

Nit: 4. Please squash the commits.

File: server/etcdserver/server.go [hunk 5]
```diff
@@ -1051,6 +1051,15 @@ func (s *EtcdServer) run() {
 					}
 					lid := lease.ID
 					s.GoAttach(func() {
+						if !s.isLeader() || !s.isLeaderInRaftStateMachine() {
```
[the remark this answers is not in the document]
Comment (jiapeish) [reply server/etcdserver/server.go:1059]: I noticed that in etcd 3.4/3.3, lg maybe nil;
In master, lg seems to be not nil...

**Comment by xiang90** [comment]
@tangcong can you push again to re-trigger travis ci?

**Comment by tangcong** [comment]
## Procfile.gofail file
```
# Use goreman to run `go get github.com/mattn/goreman`
etcd1: GOFAIL_HTTP="127.0.0.1:1111" bin/etcd --name infra1 --listen-client-urls http://127.0.0.1:2379 --advertise-client-urls http://127.0.0.1:2379 --listen-peer-urls http://127.0.0.1:12380 --initial-advertise-peer-urls http://127.0.0.1:12380 --initial-cluster-token etcd-cluster-1 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state new --enable-pprof --logger=zap --log-level=debug
etcd2: GOFAIL_HTTP="127.0.0.1:1112" bin/etcd --name infra2 --listen-client-urls http://127.0.0.1:22379 --advertise-client-urls http://127.0.0.1:22379 --listen-peer-urls http://127.0.0.1:22380 --initial-advertise-peer-urls http://127.0.0.1:22380 --initial-cluster-token etcd-cluster-1 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state new --enable-pprof --logger=zap --log-level=debug
etcd3: GOFAIL_HTTP="127.0.0.1:1113" bin/etcd --name infra3 --listen-client-urls http://127.0.0.1:32379 --advertise-client-urls http://127.0.0.1:32379 --listen-peer-urls http://127.0.0.1:32380 --initial-advertise-peer-urls http://127.0.0.1:32380 --initial-cluster-token etcd-cluster-1 --initial-cluster 'infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380' --initial-cluster-state new --enable-pprof --logger=zap --log-level=debug
```

## reproduce.sh

```
#!/bin/bash
# gofail enable,build etcd
cd ./server && go get github.com/etcd-io/gofail/runtime && cd ../ 
cd ./etcdctl && go get github.com/etcd-io/gofail && cd ../
FAILPOINTS="true" make build

# start etcd and enable the FAILPOINT HTTP server
nohup goreman -f Procfile.gofail start &

# waiting for etcd ready
members=( 2379 22379 32379)
for port in ${members[@]}
do
echo $port
echo "Waiting for etcd $port ready..."
while ! nc -z 127.0.0.1 $port; do
  sleep 1
done
done

# grant lease,put key,keep-alive
leaseid=`etcdctl lease grant 3 | awk -F" " '{ print $2}'`
echo $leaseid
etcdctl put hello world --lease $leaseid
members=( 2379 22379 32379)
for port in ${members[@]}
do
nohup etcdctl --endpoints=http://127.0.0.1:$port lease keep-alive $leaseid > lease.log 2>&1 &
done

# inject disk-io delay failure into leader
members=( 2379 22379 32379)
gofailport=1111
for port in ${members[@]}
do
leader=`curl -s 127.0.0.1:$port/metrics | grep etcd_server_is_leader | tail -1 | grep 1`
echo $port,$leader
if [ -n "$leader" ]; then
    echo "127.0.0.1:$port is leader","gofail http addr is",$gofailport
    echo "inject disk-io delay failure into leader"
    curl http://127.0.0.1:$gofailport/etcdserver/raftAfterSave -XPUT -d'sleep(4000)'
    curl http://127.0.0.1:$gofailport
    etcdctl put hello1 world2
    break
fi
gofailport=$(($gofailport+1))
done

# wait result
echo "wait for result:"
sleep 5
grep -rn leader-in-raft-state-machine nohup.out

#{"level":"warn","ts":"2021-04-23T16:21:09.897+0800","caller":"etcdserver/server.go:1091",
# "msg":"ignore older leader revoke lease request",
# "lease-id":"326978fdcba0b504","is-leader":true,"is-leader-in-raft-state-machine":false}
```
## reproduce the issue, verify the pr
```
./reproduce.sh
```
@ptabor it is hard to add integration and unit test case, I provide a simple script to inject disk-io delay failure into the leader, it can reproduce the issue verify the pr correctness.
if necessary，maybe we can add a case to simulate this problem in the functional test in another pr.

**Comment by tangcong** [comment]
> @tangcong can you push again to re-trigger travis ci?

done, travis ci is green. @xiang90 @gyuho

**Comment by xiang90** [comment]
@tangcong 

Adding it into the functional test should be good enough (translating your shell script into a Go program basically).

**Comment by jiapeish** [comment]
From a user's perspective, it would be better to solve this issue first and then put a unit-test later in an independent PR.

**Comment by tangcong** [comment]
> @tangcong
> 
> Adding it into the functional test should be good enough (translating your shell script into a Go program basically).

done. please see pr #12898. thanks.

**Comment by tangcong** [comment]
> @tangcong
> 
> Adding it into the functional test should be good enough (translating your shell script into a Go program basically).

done. please see pr #12898. thanks. @xiang90

**Comment by xiang90** [comment]
@tangcong Resolve the conflict? Also, why does the test still pass after adding https://github.com/etcd-io/etcd/pull/12898?

**Comment by tangcong** [comment]
@xiang90 [FAILPOINT TEST CASE](https://github.com/etcd-io/etcd/blob/master/tests/functional/functional.yaml#L173) is not enabled by default, I guess it may be too flaky, so my new FAILPOINT test case is also disabled.  I'm not sure if etcd release team ran FAILPOINT TEST CASE in other environments when they released the new version of etcd. There was a bug( fixed by #12898)  in it before and it seemed to be unable to run.

**Comment by tangcong** [comment]
@xiang90 @gyuho PTAL. thanks. May 17th as code freeze for the 3.5 release.

**Comment by jiapeish** [comment]
It seems that we missed the v3.5 code freeze...

**Comment by xiang90** [comment]
@tangcong @jiapeish 

If the raft routine also blocks for some time, the isLeader might still return true while a new leader is already elected?

I think it is safter to:

1. only apply lease revoke request from the current term. if a revoke is from a previous leader, just ignore it.
or 2. not forward lease revoke request to raft leader (I am not sure if this is easy to implement or not)

**Comment by xiang90** [comment]
@tangcong @jiapeish 

This is a bug fix rahter than a new future. So we do not need to worry about the releases.

**Pull request closed without merging**